### PR TITLE
wallet compatibility issue

### PR DESCRIPTION
**Pull request by oxarbitrage**
https://github.com/bitshares/bitshares-core/issues/1307

- Expose `account_id_to_string` so it can be called from inside wallet api functions, this is not reflected to clients so they can't execute it from the command line.
- Apply a fix to `get_account_balances` only as i want to make sure the solution is good enough before going forward.
- Added `Todo` comments to remove code after hardfork. Will add this to all the calls that need to be modified.

Please review so i can apply in more places. Will add commits to this same pull request, 1 per function modified.

### CODE REVIEW

**Comment by oxarbitrage** [comment]
Forgot to mention, this works in the nodes with the new version and also in older ones using account id or account_name.

File: libraries/wallet/include/graphene/wallet/wallet.hpp [hunk 1]
```diff
@@ -476,6 +476,13 @@ class wallet_api
        */
       account_object                    get_account(string account_name_or_id) const;
 
+      /** Returns string representation of an account ID.
+       *
+       * @param id the ID of the account to be converted to string
+       * @returns string representation of account ID
+       */
+      string                    account_id_to_string(account_id_type id) const;
```
Comment (abitmore) [inline libraries/wallet/include/graphene/wallet/wallet.hpp]: Since we already have a function in the impl class, I think we don't need it here.

File: libraries/wallet/wallet.cpp [hunk 2]
```diff
@@ -2961,7 +2961,15 @@ map<string,account_id_type> wallet_api::list_accounts(const string& lowerbound,
...
 vector<asset> wallet_api::list_account_balances(const string& id)
 {
-   return my->_remote_db->get_account_balances(id, flat_set<asset_id_type>());
+   /*
+    * Compatibility issue
+    * Current Date: 2018-09-13 More info: https://github.com/bitshares/bitshares-core/issues/1307
+    * Todo: remove the next 2 lines and change always_id to id in remote call after next hardfork
+    */
+   auto account = get_account(id);
+   auto always_id = account_id_to_string(account.id);
```
Comment (abitmore) [inline libraries/wallet/wallet.cpp]: We can call `my->account_id_to_string(id)` directly.

File: libraries/wallet/include/graphene/wallet/wallet.hpp (hunk 1 above)
Comment (oxarbitrage) [reply libraries/wallet/include/graphene/wallet/wallet.hpp]: silly me, fixed here https://github.com/bitshares/bitshares-core/pull/1323/commits/cab7ff4dae14d713bc21698f57f3ca41c37d585d thanks.

**Comment by oxarbitrage** [comment]
list of wallet calls to be reviewed/changed in this pull request to be more defensive until next HF:

 - [ ] resync
 - [ ] get_account
 - [ ] import_balance
 - [ ] create_committee_member
 - [ ] get_witness
 - [ ] get_committee_member
 - [ ] create_witness
 - [ ] get_vesting_balances
 - [ ] vote_for_committee_member
 - [ ] vote_for_witness
 - [x] list_account_balances
 - [x] get_account_history
 - [ ] get_account_history_by_operations
 - [ ] get_relative_account_history

**Comment by oxarbitrage** [comment]
@abitmore i was able to finish reviewing and adding code for compatibility issue here. please take a look when you get the time. thank you very much.

File: libraries/wallet/wallet.cpp [hunk 3]
```diff
@@ -2997,7 +3043,7 @@ vector<operation_detail> wallet_api::get_account_history(string name, int limit)
 
       int page_limit = skip_first_row ? std::min( 100, limit + 1 ) : std::min( 100, limit );
 
-      vector<operation_history_object> current = my->_remote_hist->get_account_history( name, operation_history_id_type(),
+      vector<operation_history_object> current = my->_remote_hist->get_account_history( always_id, operation_history_id_type(),
```
Comment (jmjatlanta) [inline libraries/wallet/wallet.cpp]: It seems Github is limiting line lengths to 123 characters without scrolling. Please adjust ( it is currently cut off at operation_history_id_ty )

File: libraries/wallet/wallet.cpp [hunk 4]
```diff
@@ -3034,14 +3080,21 @@ vector<operation_detail> wallet_api::get_relative_account_history(string name, u
...
+
    if(start == 0)
        start = stats.total_ops;
    else
       start = std::min<uint32_t>(start, stats.total_ops);
 
    while( limit > 0 )
    {
-      vector <operation_history_object> current = my->_remote_hist->get_relative_account_history(name, stop, std::min<uint32_t>(100, limit), start);
+      vector <operation_history_object> current = my->_remote_hist->get_relative_account_history(always_id, stop, std::min<uint32_t>(100, limit), start);
```
Comment (jmjatlanta) [inline libraries/wallet/wallet.cpp]: Line length. Gitub cuts it off at "always_id, stop, std::min"

**Comment by oxarbitrage** [comment]
thank you for the test @jmjatlanta

**Comment by jmjatlanta** [comment]
Isn't problems with spacing fun? Here's a good (my opinion) talk on the subject: https://www.youtube.com/watch?v=ZsHMHukIlJY

**Comment by oxarbitrage** [comment]
@jmjatlanta i am all in favor of implementing code styles practices and clean code. Another one i like is "Clean Code. A Handbook of Agile Software Craftsmanship" by Robert C. Martin. It goes in a very similar direction.

I just changed some long function calls and definitions at https://github.com/bitshares/bitshares-core/pull/1323/commits/496c6229e13bd511c2380f9c8d540e68bd65a65d
This is all arguments in new line and 6 spaces(2 tabs) indentation for them, is a method that looks clear and can scale; recommended in the video as an option.

We should discuss this and other similar at https://github.com/bitshares/bitshares-core/issues/1318
We cant follow any guide as a bible but we can take some ideas from them and write a set of rules for us.

**Comment by oxarbitrage** [comment]
One of the things all clean code guides recommended is small functions that do only one thing. This is a rule that in our codebase is pretty hard to follow, i practiced this in one of the refactor of ES plugin, they are now all very short functions that try to do only one thing, only one return path and have the less number of arguments possible.
I think that is great but hard to apply everywhere.

Apart from that there are other rules we will find hard to follow, we need to make our own specific rules.

**Comment by cogutvalera** [comment]
> One of the things all clean code guides recommended is small functions that do only one thing. This is a rule that in our codebase is pretty hard to follow, i practiced this in one of the refactor of ES plugin, they are now all very short functions that try to do only one thing, only one return path and have the less number of arguments possible.
> I think that is great but hard to apply everywhere.
> 
> Apart from that there are other rules we will find hard to follow, we need to make our own specific rules.

@oxarbitrage agree that small functions that do only one thing with as less args as possible is better than huge functions, etc. 

In general the simplest solutions are the best always ! At the same time it isn't so easy to find the simplest solutions always :)

**Comment by oxarbitrage** [comment]
Yea, the brackets thing is just what i wrote here https://github.com/bitshares/bitshares-core/issues/1318#issuecomment-429999963

To me they are all the same, however if we want to make some style guide we should select 1 form(or more, or allow all) and use it the majority of the times as you said. I am ok with that, don't need to be a bible.
